deviceplugin/usb.go

```diff
@@ -69,54 +73,85 @@ type usbDevice struct {
 }
 
+// queryUSBDeviceCharacteristicsByDirectory scans the given directory for information regarding the given USB device,
+// then returns a pointer to a new usbDevice if information is found.
+func queryUSBDeviceCharacteristicsByDirectory(dir os.DirEntry) (result *usbDevice, err error) {
+	if !dir.IsDir() {
+		// There shouldn't be any raw files in this directory, but just in case.
+		return
+	}
+	// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
+	vnd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirVendorIDFile)
+	if err != nil {
+		// We can't read the vendor file for some reason, it probably doesn't exist.
+		return
+	}
+	prd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirProductIDFile)
+	if err != nil {
+		return
+	}
+
+	// The following two calls shouldn't fail.
+	busRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusFile)
+	if err != nil {
+		return result, err
+	}
+	bus := binary.LittleEndian.Uint16(busRaw)
+	busLocRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusDevFile)
+	if err != nil {
+		return result, err
+	}
+	busLoc := binary.LittleEndian.Uint16(busLocRaw)
+
+	res := usbDevice{
+		Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
+		Product:   usbID(binary.LittleEndian.Uint16(prd)),
+		Bus:       bus,
+		BusDevice: busLoc,
+	}
+	return &res, nil
+}
+
+// enumerateUSBDevices rapidly scans the OS system bus for attached USB devices.
+// Pure Go; does not require external linking.
 func enumerateUSBDevices() (specs []usbDevice, err error) {
 	allDevs, err := os.ReadDir(usbDevicesDir)
 	if err != nil {
 		return
 	}
-	for _, dev := range allDevs {
-		if !dev.IsDir() {
-			// There shouldn't be any raw files in this directory, but just in case.
-			continue
-		}
-		// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
-		vnd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirVendorIDFile)
-		if err != nil {
-			// We can't read the vendor file for some reason, it probably doesn't exist.
-			continue
-		}
-		prd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirProductIDFile)
-		if err != nil {
-			continue
-		}
 
-		// The following two calls shouldn't fail.
-		busRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusFile)
-		if err != nil {
-			return specs, err
-		}
-		bus := binary.LittleEndian.Uint16(busRaw)
-		busLocRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusDevFile)
-		if err != nil {
-			return specs, err
-		}
-		busLoc := binary.LittleEndian.Uint16(busLocRaw)
+	var wg sync.WaitGroup
+	for i, dev := range allDevs {
+		wg.Add(1)
 
-		spec := usbDevice{
-			Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
-			Product:   usbID(binary.LittleEndian.Uint16(prd)),
-			Bus:       bus,
-			BusDevice: busLoc,
-		}
-		specs = append(specs, spec)
+		// Copy the loop variables
+		index := i
```

Comment:
you can even do `i:= i`

Reply:
.... my linter was complaining at me 5 minutes ago and now isn't, fixed in b804deb